src/Workspaces/Remote/ServiceHub/Shared/ClientDirectStream.cs

```diff
@@ -3,17 +3,64 @@
 #nullable enable
 
 using System;
+using System.Diagnostics;
 using System.IO;
 using System.IO.Pipes;
 using System.Runtime.Remoting;
 using System.Threading;
 using System.Threading.Tasks;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
-    internal static class DirectStream
+    /// <summary>
+    /// Direct stream between server and client to pass around big chunk of data
+    /// </summary>
+    internal sealed class ClientDirectStream : Stream
     {
-        public static async Task<Stream> GetAsync(string streamName, CancellationToken cancellationToken)
+        // 4KB buffer size
+        private const int BufferSize = 4 * 1024;
+        private const int ConnectWithoutTimeout = 1;
+        private const int MaxRetryAttemptsForFileNotFoundException = 3;
+        private const int ErrorSemTimeoutHResult = unchecked((int)0x80070079);
+
+        private static readonly TimeSpan s_connectRetryInterval = TimeSpan.FromMilliseconds(20);
+
+        private readonly string _name;
+        private readonly NamedPipeClientStream _pipe;
+        private readonly Stream _stream;
+
+        public ClientDirectStream(string name)
+        {
+            // this type exists so that consumer doesn't need to care about all these arguments/flags to get good performance
+            _name = name;
+            _pipe = new NamedPipeClientStream(serverName: ".", pipeName: name, direction: PipeDirection.Out);
+            _stream = new BufferedStream(_pipe, BufferSize);
+        }
+
+        public string Name => _name;
+
+        public static async Task WriteDataAsync<TData>(string streamName, TData data, Func<ObjectWriter, TData, CancellationToken, Task> dataWriter, CancellationToken cancellationToken)
+        {
+            try
+            {
+                using var stream = await CreateStreamAsync(streamName, cancellationToken).ConfigureAwait(false);
+
+                using (var objectWriter = new ObjectWriter(stream, cancellationToken))
+                {
+                    await dataWriter(objectWriter, data, cancellationToken).ConfigureAwait(false);
+                }
+
+                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
does the stream not flush on disposal?  we might want to consider havin it do that to prevent accidents.